In [1]:
import sys
from time import time
from email_preprocess import preprocess
from visualize import plotIt
from sklearn.svm import SVC
%matplotlib inline

In [2]:
features_train, features_test, labels_train, labels_test = preprocess()

no. of Chris training emails: 7936
no. of Sara training emails: 7884


In [3]:
clf = SVC(kernel="linear")

In [4]:
t0 = time()
clf.fit(features_train,labels_train)
print ("training time for SVM:", round(time()-t0, 3), "s")

training time for SVM: 158.881 s


Training time for Naive Bayes for same data was 1.127 s

In [5]:
t1 = time()
pred = clf.predict(features_test)
print ("testing time for SVM:", round(time()-t1, 3), "s")

testing time for SVM: 16.738 s


Testing time for Naive Bayes for same data was 0.159 s

In [6]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(pred, labels_test)

print (acc*100)45

98.40728100113766


Accuracy of Naive Bayes Classifier for same dataset : 97.32650739476678

# Applying Speed-Accuracy-TradeOff

#### slicing the training dataset to 1/100th

In [7]:
features_train_sliced = features_train[:int(len(features_train)/100)] 
labels_train_sliced = labels_train[:int(len(labels_train)/100)] 

In [8]:
t0 = time()
clf.fit(features_train_sliced,labels_train_sliced)
print ("After Speed-Accuracy-TradeOff training time for SVM:", round(time()-t0, 3), "s")

After Speed-Accuracy-TradeOff training time for SVM: 0.116 s


In [9]:
t1 = time()
pred = clf.predict(features_test)
print ("After Speed-Accuracy-TradeOff testing time for SVM:", round(time()-t1, 3), "s")

After Speed-Accuracy-TradeOff testing time for SVM: 1.006 s


In [10]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(pred, labels_test)

print ("After applying Speed-Accuracy-TradeOff accuracy has reduced to: ")
print (acc*100)

After applying Speed-Accuracy-TradeOff accuracy has reduced to: 
88.45278725824801


# Checking the accuracy w Speed-Accuracy-Tradeoff and rbf kernel

In [11]:
clf = SVC(kernel="rbf")

In [12]:
t0 = time()
clf.fit(features_train_sliced,labels_train_sliced)
print ("After rbf kernel and Speed-Accuracy-TradeOff training time for SVM:", round(time()-t0, 3), "s")

After rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.124 s


In [13]:
t1 = time()
pred = clf.predict(features_test)
print ("After rbf kernel and Speed-Accuracy-TradeOff testing time for SVM:", round(time()-t1, 3), "s")

After rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 1.153 s


In [14]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(pred, labels_test)

print ("After rbf kernel and applying Speed-Accuracy-TradeOff accuracy has reduced to: ")
print (acc*100)

After rbf kernel and applying Speed-Accuracy-TradeOff accuracy has reduced to: 
61.60409556313993


# Checking the accuracy w Speed-Accuracy-Tradeoff, rbf kernel, and optimizing param C

###### After various iterations, the optimal value of C comes out to be 10000.0 where accuracy increases to about 89%

In [15]:
clf = SVC(kernel="rbf", C=10000.0)

In [16]:
t0 = time()
clf.fit(features_train_sliced,labels_train_sliced)
print ("After optimizing C, rbf kernel and Speed-Accuracy-TradeOff training time for SVM:", round(time()-t0, 3), "s")

After optimizing C, rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.104 s


###### After optimizing C=10.0, rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.124 s
###### After optimizing C=100.0, rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.124 s
###### After optimizing C=1000.0, rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.134 s
###### After optimizing C=10000.0, rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.12 s
###### After optimizing C=100000.0, rbf kernel and Speed-Accuracy-TradeOff training time for SVM: 0.117 s

In [17]:
t1 = time()
pred = clf.predict(features_test)
print ("After optimizing C, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM:", round(time()-t1, 3), "s")

After optimizing C, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 0.909 s


###### After optimizing C=10.0, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 1.137 s
###### After optimizing C=100.0, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 1.183 s
###### After optimizing C=1000.0, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 1.153 s
###### After optimizing C=10000.0, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 0.932 s
###### After optimizing C=100000.0, rbf kernel and Speed-Accuracy-TradeOff testing time for SVM: 0.886 s

In [18]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(pred, labels_test)

print ("After optimizing C, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: ")
print (acc*100)

After optimizing C, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: 
89.24914675767917


###### After optimizing C=10.0, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: 61.60409556313993
###### After optimizing C=100.0, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: 61.60409556313993
###### After optimizing C=1000.0, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: 82.13879408418657
###### After optimizing C=10000.0, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: 89.24914675767917
###### After optimizing C=100000.0, rbf kernel and applying Speed-Accuracy-TradeOff accuracy is: 86.00682593856655

# Checking the accuracy without Speed-Accuracy-Tradeoff, rbf kernel, and optimizing param C 

In [19]:
clf = SVC(kernel="rbf", C=10000.0)

In [20]:
t0 = time()
clf.fit(features_train,labels_train)
print ("After optimizing C, rbf kernel for full dataset training time for SVM:", round(time()-t0, 3), "s")

After optimizing C, rbf kernel for full dataset training time for SVM: 101.676 s


In [21]:
t1 = time()
pred = clf.predict(features_test)
print ("After optimizing C, rbf kernel for full dataset testing time for SVM:", round(time()-t1, 3), "s")

After optimizing C, rbf kernel for full dataset testing time for SVM: 10.591 s


In [22]:
from sklearn.metrics import accuracy_score
acc = accuracy_score(pred, labels_test)

print ("After optimizing C, rbf kernel for full dataset accuracy is: ")
print (acc*100)

After optimizing C, rbf kernel for full dataset accuracy is: 
99.08987485779295
